work on embedded

cheqianh opened the pull request:
This pull request is based on pr1: https://github.com/cheqianh/ion-java/pull/1/files

What I did:
1. --output-format none
2. error report for reading 
3. embedded stream 
4. combined some functions to make code simpler 

Questions:
1. There are 3 error types, READ happens when reading ionStream to EventStream and WRITE happens when we read EventStream to ionStream right? what is STATE?

2. For constant, we use capital letters with underscore. Can I use this style for global variable? (for example, variable in inProcess.java 48-49)

3. When I read 2 files, there is an IVM between the EventStream of two files. ion-c doesn't have that.
For example:

my command is: ion-java process -f events file1 file2 and there is a IVM for the output
output:
{                                                                       
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;eventType: ...                                                                 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ionType: ...                                                                      
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;...                                                                                     
}     
......                                                                                    
{                                                                                         
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;eventType: STREAM_END                                             
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;depth: 0                                                                          
}                                                                         
**$ion_1_0**
output:
{                                                                       
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;eventType: ...                                                                 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ionType: ...                                                                      
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;...                                                                                     
}       
......                                                                                  
{                                                                                         
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;eventType: STREAM_END                                             
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;depth: 0                                                                          
}

cheqianh committed:
work on embedded

done embedded stream

finished none type

finish error report, start working on re-write event type

added embedded stream for pretty/text format and changed function's order preparing for re-write part

ready for cr

cheqianh committed:
work on re-write

tgregg commented:
In response to your questions:
1. The `STATE` error type is basically for IllegalState-type errors. Things that are never expected to happen that aren't necessarily directly related to reading or writing Ion.
2. All caps with underscores is used in Java style for constants (i.e. enum values or variables labeled `static final`). It should not be used for any other type of variable. In general, global variables that are not constants (i.e. `static` but not `final`) should be avoided at pretty much all costs. For example, I'd consider how you can avoid it on lines 48-49.
3. That's fine. It looks like ion-java has different default behavior around when it writes text IVMs. Unlike the binary format, where the IVM is always required, it is usually optional in text.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -37,11 +40,16 @@
     private static final String SYSTEM_OUT_DEFAULT_VALUE = "out";
     private static final String SYSTEM_ERR_DEFAULT_VALUE = "err";
+    private static final String EMBEDDED_STREAM_ANNOTATION = "embedded_documents";
+    private static String CURRENT_FILE = null;
+    private static int EVENT_INDEX = 1;
```
Try not to use non-constant globals.

cheqianh replied:
Used a class which stores current file's information (ionProcess.java 85)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -37,11 +40,16 @@
+    private static int EVENT_INDEX = 1;
 
     public static void main(final String[] args) {
+
```
Superfluous blank line.

cheqianh replied:
Removed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -67,118 +75,169 @@ public static void main(final String[] args) {
-            IonReader ionReader = IonReaderBuilder.standard().build(inputStream);
+            try {
+                CURRENT_FILE = path;
+                InputStream inputStream = new BufferedInputStream(new FileInputStream(path));
+                IonReader ionReader = IonReaderBuilder.standard().build(inputStream);
```
The inputStream and ionReader need to be closed when done, so consider making this a try-with-resources block.

cheqianh replied:
Used try-with-resources blocks for all temporary ionReader/ionWriter (ionProcess.java 88-91) (ionProcess.java 157 - 161) (ionProcess.java 219) (ionProcess.java 364-366) (ionProcess.java 635).

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -277,6 +336,26 @@ private static boolean isSameSymbolTable(SymbolTable x, SymbolTable y) {
+    private static boolean isEmbeddedStream(IonReader ionReader) {
+        return (ionReader.getType() == IonType.SEXP || ionReader.getType() == IonType.LIST)
+                && ionReader.getDepth() == 0
+                && ionReader.getTypeAnnotations() != null
+                && ionReader.getTypeAnnotations().length > 0
+                && ionReader.getTypeAnnotations()[0].equals(EMBEDDED_STREAM_ANNOTATION);
```
`IonReader.getTypeAnnotations()` creates a new array on each invocation, so it's wasteful to call it multiple times while positioned at the same value. Consider storing the result in a variable.

Additionally, the method is documented not to return null, so the null check isn't necessary.

cheqianh replied:
Changed. (IonProcess.java 699)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -67,118 +75,169 @@ public static void main(final String[] args) {
+                    String stream = ionReader.stringValue();
+                    IonReader tempIonReader = IonReaderBuilder.standard().build(stream);
+                    ByteArrayOutputStream text = new ByteArrayOutputStream();
+                    IonWriter tempIonWriter = IonTextWriterBuilder.standard().build(text);
```
Use the same IonWriterBuilder you used to create `ionWriterForOutput` to guarantee they have the same configuration.

cheqianh replied:
I tested an example on ion-c, 
For example, **the file is:**
embedded_documents::(
'''$ion_1_0 abc 12 abc'''
)

**and I run the commands respectively:** 
./ion process -f pretty fileName 
and 
./ion process -f text fileName

**I got the same output for both commands:** 
embedded_documents::(
  "abc 12 abc"
)

I tested the same file with command **./ion process -f pretty fileName**  in ion-java, it shows:
embedded_documents::(
  "\nabc\n12\nabc"
)

So I think for text format(both pretty and text format) we use IonTextWriterBuilder.**standard()**.build(text); and binary format we use IonTextWriterBuilder.**binary()**.build(text), right?

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -67,118 +75,169 @@ public static void main(final String[] args) {
+                    ByteArrayOutputStream text = new ByteArrayOutputStream();
+                    IonWriter tempIonWriter = IonTextWriterBuilder.standard().build(text);
+
+                    //read each string
```
Inaccurate comment: the values are not necessarily all strings.

cheqianh replied:
deleted

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -67,118 +75,169 @@ public static void main(final String[] args) {
+                    }
+
+                    //close ionReader and ionWriter
+                    ionWriterForOutput.writeString(text.toString());
```
Always specify the charset when using byte-to-string APIs like `ByteArrayOutputStream.toString`. ("UTF-8" is the right choice.)

Another option is to provide the `IonTextWriterBuilder` with a sink that goes directly to String instead of using `ByteArrayOutputStream`. Try `StringBuilder`, which is an `Appendable`.

cheqianh replied:
I used String build to append bytes and convert it to string by tostring() method at the end. (IonProcess.java 166)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -67,118 +75,169 @@ public static void main(final String[] args) {
+                    IonReader tempIonReader = IonReaderBuilder.standard().build(stream);
+
+                    while (tempIonReader.next() != null) {
+                        System.out.println("\nprinting: tempIonReader.getType: ");
+                        System.out.println(tempIonReader.getType());
```
Leftover debug printlns

cheqianh replied:
Removed, sorry about that.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -67,118 +75,169 @@ public static void main(final String[] args) {
+                    while (tempIonReader.next() != null) {
+                        System.out.println("\nprinting: tempIonReader.getType: ");
+                        System.out.println(tempIonReader.getType());
+                        EVENT_INDEX = ionStreamToEvent(EventType.SCALAR, tempIonReader)
```
The value might not be a scalar; embedded streams can contain values of any type. You may need a recursive call here.

cheqianh replied:
Changed. (IonProcess.java 221)

tgregg commented on test/tools/events/Event.java:
```diff
@@ -28,70 +33,86 @@ public Event(EventType eventType, IonType ionType, SymbolToken fieldName, Symbol
+        //event index + 1 if we write OutputStream successfully.
+        return eventIndex + 1;
     }
 }
```
For fun, let's end every file with a newline.

cheqianh replied:
Sure added.

tgregg commented on test/tools/cli/OutputFormat.java:
```diff
@@ -42,6 +42,12 @@ public IonWriter createIonWriter(OutputStream outputStream) {
+    NONE {
+        @Override
+        public IonWriter createIonWriter(OutputStream outputStream) {
+            return null;
```
This should be an IonWriter that writes to nowhere. We still want to exercise all the logic of an IonWriter; we just don't want the data to be written anywhere. You can make a trivial `OutputStream` implementation to achieve this.

cheqianh replied:
Created a trivial OutputStream for none type. (IonProcess.java 245-247) 

I made a help function that returns an outputStream (IonProcess.java 241). Even though I used try-with-resource block to close the returned outputStream, do I need to close the trivial outputStream within the help function I made? (ionProcess.java 246)

cheqianh committed:
changed global variable to class

cheqianh committed:
finish try-with-resource anootation check charset

cheqianh committed:
fix embedded stream bugs

cheqianh committed:
finished embedded stream

cheqianh committed:
finished none type and fix the exception bugs

cheqianh committed:
ready for cr

cheqianh committed:
fix cr errors

cheqianh committed:
change function names for rewrite

finished ion To Event struct

work on rewrite

work on embedded event to ion

finish rewrite embeddedStream

fixing bugs

work on etra "

finish string debug

almost done rewrite

finish rewrite working on style

checking styles for cr

working on /bin/zsh

ready for cr

finished pr check

cheqianh commented:
The new commit fixes the above issues and implement rewriting EventStream to IonStream. 

**Description/Overall logic**  
Checking if the file starts with $ion_event_stream (ionProcess.java 95)
1.  if no, go to the logic that I wrote earlier which is pr1
2. if yes, check the desire format (--output--format -f)

    2.1.   if -f is events.  go to processToEventStreamFromEventStream(ionProcess.java 271)
    2.2.  all else go to processToNormalFromEventStream(ionProcess.java 274)

For both 2.1 and 2.2, I step in each struct of the input file and convert each struct to Event structure first, then validate if the event is valid (comparing text_value and binary_value, checking if it misses EVENT_TYPE etc.). After that, I use the event structure to write ionStream/EventStream to outputStream.

**Test**
1. simple eventStream.
2. eventStream that has STATE error.
2. eventStream that has WRITE error.
3. complex eventStream that include SEXP, LIST, BOOL, SYMBOL and so on.
2. eventStream with embedded Stream.
3. eventStream with embedded Stream that includes embedded Stream.

**Question/Unsure area**
1. When I read a string value: "some_string" by ionReader, it represents like "some_string". Then I write this string to outputStream, It is represented as "\\"some_string\\"" . I used substring method to remove extra " for writeString to match ion-c's performance, is it right? (IonProcess.java 450)

2. Since SymbolToken is an interface. Can I use fakeSymbolToken to setting filed_name and annotations? (IonProcess.java 489, 510)

3. For symbol_table event, how can I use ionWriter to create a new local table?
I left the TODO comment here (IonProcess.java 335 - 340) (IonProcess.java 395)

4. To write a text_value to timeStamp, can I use ionWriter.writeTimestamp(Timestamp.valueOf(text_value))?
(ionProcess.java 443)

5. It shows extra ' in text when I write $0 in ion-java. (ie. it shows '$0' instead of $0) 
I tried to set fieldName to $0 like ion-c when field is null or field text is empty, but it showed '$0' instead of $0. (IonProcess.java 418-425)

6. when I convert a eventStream to embedded Stream, the double quotation marks will become \\\" or \\\\\" where the number of slash is depends on the depth.  
For example, embedded_documents::("10" "embedded_documents::(\\\"10\\\" &nbsp;\\\"embedded_documents::(\\\\\\\"10\\\\\\\")\\\")")
ion-c is able to convert this embedded stream back to event stream so I think it performances correctly, but ion-c doesn't have these slashes when I convert a eventStream to embedded stream above. Did I miss anything?


Thanks!

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -45,136 +46,162 @@
                 IonWriter ionWriterForOutput = outputFormat.createIonWriter(outputStream);
                 IonWriter ionWriterForErrorReport = outputFormat.createIonWriter(errorReportOutputStream);
-        ) {
+                ) {
```
Personally I prefer the previous indentation style.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/CurrentInfo.java:
```diff
@@ -0,0 +1,40 @@
+
+import tools.events.EventType;
+
+public class CurrentInfo {
```
Maybe a name like `ProcessContext` is more accurate?

cheqianh replied:
Sure.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -45,136 +46,162 @@
                             .writeOutput(ionWriterForErrorReport);
+                }
+                System.exit(IO_ERROR_EXIT_CODE);
+            }  catch (IllegalStateException e) {
```
I think `IllegalStateException` may be too specific here. I think it should be any `Exception` (since IonException was already handled above).

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -45,136 +46,162 @@
                             .writeOutput(ionWriterForErrorReport);
                 }
                 System.exit(IO_ERROR_EXIT_CODE);
+            } catch (FileNotFoundException e) {
```
FileNotFoundException is an IOException, so it can be handled above. No need for this branch.

cheqianh replied:
Changed.

cheqianh committed:
fix writing bug

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -45,136 +46,162 @@
+                            tempIonWriter.writeValue(tempIonReader);
+                        }
+
+                        StringBuilder stringBuilder = new StringBuilder();
+                        stringBuilder.append(text);
```
I think I was confusing in my previous comment. It's not necessary to have `tempIonWriter` use the configured output format. It should always be text.

So if you have `IonWriter tempIonWriter = IonTextWriterBuilder.standard().build(text);`, there's no need to have a `ByteArrayOutputStream` at all, because that `build` method will accept a `StringBuilder`.

In the current revision, you're appending `text` to your stringBuilder, which implicitly calls `text.toString()`, which isn't what we want.

cheqianh replied:
Changed. (IonProcess.java 163)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                //update eventIndex
+                currentInfo.setEventIndex(currentInfo.getEventIndex() + 1);
+            }
+        } catch (IonException | NumberFormatException | IllegalStateException e) {
```
Can these just be thrown all the way up the call stack to where they're handled in the uppermost method?

cheqianh replied:
Changed.
The reason I left them here is because that I tried to separate WRITE and READ types. Now I add a new variable to ProcessContext to separate them.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+            if (currentInfo.getLastEventType() != EventType.STREAM_END) {
+                throw new IonException("EventStream doesn't end with STREAM_END event");
+            }
+        } catch (IonException | NumberFormatException | IllegalStateException e) {
```
Same question about letting these be thrown up the call stack.

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                        embeddedEventToIon(ionReader, ionWriterForOutput, currentInfo, args);
+                    } else {
+                        IonType type = event.getIonType();
+                        String fieldName = event.getFieldName() == null ? null : event.getFieldName().getText();
+                        SymbolToken[] annotations = event.getAnnotations();
+
+                        if (fieldName != null && fieldName.length() != 0) {
+                            ionWriterForOutput.setFieldName(fieldName);
+                        }
+                        if (annotations != null && annotations.length != 0) {
+                            ionWriterForOutput.setTypeAnnotationSymbols(annotations);
+                        }
```
This logic can be applied to both CONTAINER_START and SCALAR events.

cheqianh replied:
created a function (setAnnotationAndField()) for this and use it for both events (IonProcess.java 342)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                } else if (event.getEventType().equals(EventType.SCALAR)) {
+                    writeIonByType(event, ionWriterForOutput);
+                } else if (event.getEventType().equals(EventType.SYMBOL_TABLE)) {
+                    //TODO symbol table
+                    // For SYMBOL_TABLE events, flush the writer's existing local symbol table and any buffered data,
+                    // forcing the writer to create a new local symbol table that imports the list of symbol tables
+                    // declared by the imports field of the Event. This ensures that symbol tokens with unknown text
+                    // that occur in subsequent events in the stream can be written correctly
```
In IonJava, the only way to do this is to `finish()/close()` the existing writer and create a new one that writes to the same destination. When you create the new writer, you provide the list of imports specified by this SYMBOL_TABLE event.

cheqianh replied:
added. Can I use **IonSystemBuilder.standard().build().newSharedSymbolTable(name, version, null)** to create a symbol table? (ionProcess.java 328-329)

tgregg replied:
Yes. Make sure you only construct the IonSystem once, though.

```java
private static final IonSystem ION_SYSTEM = IonSystemBuilder.standard().build();
```

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -337,11 +696,29 @@ private static boolean isSameSymbolTable(SymbolTable x, SymbolTable y) {
+    private static boolean isEventStream(IonReader ionReader) {
+        return ionReader.next() != null
+                && ionReader.getType() == IonType.SYMBOL
+                && ionReader.symbolValue().getText() != null
+                && ionReader.symbolValue().getText().equals(EVENT_STREAM);
```
if you use `EVENT_STREAM.equals(ionReader.symbolValue.getText())` you can make this null-safe in one line.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -337,11 +696,29 @@ private static boolean isSameSymbolTable(SymbolTable x, SymbolTable y) {
+    private static boolean isEmbeddedEvent(Event event) {
+        SymbolToken[] annotations = event.getAnnotations();
+        return event.getEventType() == EventType.CONTAINER_START
+                && event.getIonType() == IonType.SEXP
```
It can be SEXP or LIST.

cheqianh replied:
Changed.

tgregg commented on test/tools/events/Event.java:
```diff
@@ -12,14 +16,25 @@
-    private final byte[] valueBinary;
-    private final ImportDescriptor[] imports;
-    private final int depth;
+    private EventType eventType;
+    private IonType ionType;
+    private SymbolToken fieldName;
+    private SymbolToken[] annotations;
+    private String valueText;
+    private byte[] valueBinary;
+    private ImportDescriptor[] imports;
+    private int depth;
```
What was the reason behind making all these fields mutable?

cheqianh replied:
I create a new constructor which accepts null value so I can set each field after declaring event class (i.e IonProcess.java 432).

tgregg commented on test/tools/events/Event.java:
```diff
@@ -33,8 +48,100 @@ public Event(EventType eventType, IonType ionType, SymbolToken fieldName, Symbol
 
+    public void validate() throws IonException, IOException {
+        if (this.eventType == null) throw new IonException("event_type can't be null");
+        else if (this.ionType == null && this.eventType != EventType.STREAM_END)
```
SYMBOL_TABLE events don't need Ion types, do they? What about CONTAINER_END events?

cheqianh replied:
Added.

I noticed that ion-c's CONTAINER_END event has ion_type but it doesn't affect processing. 
For example I change CONTAINER_END event's ionType from a struct to list, it will still write the same output.

tgregg commented on test/tools/events/Event.java:
```diff
@@ -33,8 +48,100 @@ public Event(EventType eventType, IonType ionType, SymbolToken fieldName, Symbol
+            case SCALAR:
+                String textValue = this.getValueText();
+                byte[] binaryValue = this.getValueBinary();
+                if (textValue == null && binaryValue == null) {
```
This isn't allowed, is it? Doesn't every scalar event need both value_text and value_binary?

cheqianh replied:
Sure, 

question: 

where are eventStream files come from? if they just come from ion-java-cli then they will always follow our writing logic. If they can from anywhere, for example, I type an eventStream and delete its field_name/text_value on purpose. Do we need to think about these cases?

tgregg replied:
They can come from anywhere. We should validate that the serialized event stream has all the fields required. I don't think it's too complicated.

cheqianh replied:
Wrote a new validate function for it. (IonProcess.java 528)

tgregg commented on test/tools/events/Event.java:
```diff
@@ -33,8 +48,100 @@ public Event(EventType eventType, IonType ionType, SymbolToken fieldName, Symbol
+                byte[] binaryValue = this.getValueBinary();
+                if (textValue == null && binaryValue == null) {
+                } else if (textValue != null && binaryValue != null) {
+                    try (
+                            IonReader ionReaderX = IonReaderBuilder.standard().build(textValue);
+                            IonReader ionReaderY = IonReaderBuilder.standard().build(binaryValue);
+                    ) {
+                        ionReaderX.next();
+                        ionReaderY.next();
+
+                        if (!validateTwoIonReaderValue(ionReaderX, ionReaderY)) {
+                            throw new IonException("invalid Event: Text value and Binary value are different");
+                        }
+                    }
+                } else {
+                    throw new IonException("invalid Event: Text value and Binary value are different");
+                }
```
In this case I actually recommend using `IonSystem.singleValue` to load both `textValue` and `binaryValue` into `IonValue`s. Then use `Equivalence.ionValueEquals` to compare them. Then you can delete your `validateTwoIonReaderValue` method.

cheqianh replied:
Changed, ionProcess.java 408-518

tgregg commented on test/tools/events/ImportDescriptor.java:
```diff
@@ -3,25 +3,43 @@
-    private final String importName;
-    private final int maxId;
-    private final int version;
+    private String importName;
+    private int maxId;
+    private int version;
```
Do these need to be mutable?

cheqianh replied:
I set them in IonProces.java 534-540. If you think it's not necessary, I'll change them to final.

tgregg replied:
You should be able to make them final. Always prefer to make fields final. If a field isn't final, there should be a really good reason for that.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/OutputFormat.java:
```diff
@@ -37,16 +37,22 @@ public IonWriter createIonWriter(OutputStream outputStream) {
         @Override
         public IonWriter createIonWriter(OutputStream outputStream) {
-            return null;
+            return IonTextWriterBuilder.pretty().build(outputStream);
```
Maybe the no-op OutputStream implementation instance should be created here and the `outputStream` parameter ignored?

cheqianh replied:
added (OutputFormat.java 80)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+        switch (defaultValue) {
+            case SYSTEM_OUT_DEFAULT_VALUE:
+                if (args.getOutputFormat() == OutputFormat.NONE) {
+                    try (ByteArrayOutputStream trivialOut = new ByteArrayOutputStream()) {
+                        outputStream = new BufferedOutputStream(trivialOut, BUFFER_SIZE);
+                    }
```
I recommend just creating a new class called something like `NoOpOutputStream` that extends `OutputStream` but just does nothing for any of `OutputStream`'s `write` methods. It looks like this current solution will still cause data to be buffered, which is not necessary.

cheqianh replied:
created a new file for NoOpOutputStream. (NoOpOutputStream.java)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+    }
+
+    private static ImportDescriptor[] ionStreamToImportDescriptors(IonReader ionReader) {
+        LinkedList<ImportDescriptor> imports = new LinkedList<>();
```
Out of curiosity, why did you choose `LinkedList`?

cheqianh replied:
I used LinkedList a lot earlier 
I changed it to ArrayList now. which List do you think is better for this situation?

tgregg replied:
Probably an ArrayList.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+            ionReader.stepOut();
+            imports.add(table);
+        }
+        ImportDescriptor[] importsArray = new ImportDescriptor[imports.size()];
+        for (int i = 0; i < imports.size(); i++) {
+            importsArray[i] = imports.get(i);
+        }
```
Can you just use `List.toArray`?

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                                break;
+                        }
+                    }
+                    FakeSymbolToken symbolToken = new FakeSymbolToken(fieldText, fieldSid);
```
Use
```java
SymbolToken symbolToken = _Private_Utils.newSymbolToken(fieldText, fieldSid);
```

cheqianh replied:
Changed,

If I use _Private_Utils.newSymbolToken(fieldText, fieldSid); the FieldText can't be null.
For example, there is an event whose fieldName is empty. For Ion-c it will print $0 but for ion-java, we can't create a null text symbolToken so it will throw an error. Is that expected?

tgregg replied:
I don't see anything in `_Private_Utils.newSymbolToken(fieldText, fieldSid)` that prevents you from providing `null` fieldText. Where did you find that?

cheqianh replied:
Got it, fieldText must be a string rather than null value.

For example _Private_Utils.newSymbolToken(null, 0) is invalid but 
For example _Private_Utils.newSymbolToken(a_null_string, 0)  is valid.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                    ionReader.stepOut();
+                    break;
+                case "annotations":
+                    LinkedList<SymbolToken> annotations = new LinkedList<>();
```
Why a `LinkedList`?

cheqianh replied:
same comment

cheqianh replied:
Checked.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                        ionReader.stepOut();
+                    }
+                    int size = annotations.size();
+                    SymbolToken[] annotationsArray = new SymbolToken[size];
+                    for (int i = 0; i < size; i++) {
+                        annotationsArray[i] = annotations.get(i);
+                    }
```
Use `List.toArray`.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                    }
+                    event.setValueBinary(binary);
+                    ionReader.stepOut();
+                    break;
```
Your `Event` class doesn't have to actually store both value_text and value_binary. It just needs to verify they're equivalent, then store one copy of the value. It may be easiest to store the value as an `IonValue` for comparisons. This will require a little bit of a refactor here and in `Event.validate`.

cheqianh replied:
changed.

My event is only store one IonValue now.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+        while (ionReader.next() != null) {
+            try (
+                    ByteArrayOutputStream out = new ByteArrayOutputStream();
+                    IonWriter tempWriter = args.getOutputFormat().createIonWriter(out);
```
This should always be a new text writer, and you can just provide it a StringBuilder.

cheqianh replied:
Changed IonProcess.java 366

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+            if (ionWriter.isInStruct()) ionWriter.setFieldName("$0");
+        }
+
+        switch (ionType) {
```
See my other comment about just storing one copy of scalar values as an `IonValue` in `Event`. That way you don't need to decide whether to write from `valueText` or `valueBinary` and you don't need to re-parse the value.

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                ionWriter.setFieldName("$0");
+            }
+        } else {
+            if (ionWriter.isInStruct()) ionWriter.setFieldName("$0");
```
This should just be an error, right? If the writer is in a struct, there *must* be a field name.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                ionWriter.setFieldNameSymbol(field);
+            } else {
+                ionWriter.setFieldName("$0");
+            }
```
Just replace this with `ionWriter.setFieldNameSymbol(field)`, which handles the case where the symbol token has unknown text.

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -183,79 +210,411 @@ private static void processEvent(IonWriter ionWriterForOutput,
+                    currentInfo.setLastEventType(event.getEventType());
+
+                    if (isEmbeddedEvent(event)) {
+                        embeddedEventToIon(ionReader, tempWriter, currentInfo, args);
```
So this is the case where an embedded event has an embedded event? I don't think we should support that.

cheqianh replied:
Deleted

tgregg reviewed (commented):
Good progress.

I think addressing my comments on this revision will answer the questions you listed. Let me know if anything requires further clarification, though.

cheqianh committed:
finish indentation style

cheqianh committed:
change currentInfo to processContext

cheqianh committed:
still have refactor and symbol table part

cheqianh committed:
finished refactor event

cheqianh committed:
finish symbol table case

cheqianh committed:
fix symbol table bug

cheqianh committed:
checked logic

cheqianh committed:
ready for cr

cheqianh committed:
finished 'only one ionsystem'

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -61,66 +62,68 @@ public static void main(final String[] args) {
                 } else {
-                    new ErrorDescription(ErrorType.READ, e.getMessage(), currentInfo.getFileName())
+                    new ErrorDescription(processContext.getState(), e.getMessage(), processContext.getFileName())
                             .writeOutput(ionWriterForErrorReport);
```
Looks like these two branches only differ by whether the event index is provided. Can you unify the branches and handle the presence/absence of an event index within a single ErrorDescription constructor?

cheqianh replied:
Changed.

used initial value -1. (IonProcess.java 67). the writeOutput method will ignore eventIndex if it's -1.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -61,66 +62,68 @@ public static void main(final String[] args) {
-                    new ErrorDescription(ErrorType.STATE, e.getMessage(), currentInfo.getFileName())
+                    new ErrorDescription(ErrorType.STATE, e.getMessage(), processContext.getFileName())
                             .writeOutput(ionWriterForErrorReport);
                 }
```
Same question about unifying the branches.

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -265,115 +266,111 @@ private static void processToEventFromIonStream(IonWriter ionWriterForOutput,
+            //update eventIndex
+            processContext.setEventIndex(processContext.getEventIndex() + 1);
+            Event event = eventStreamToEvent(ionReader);
+            //event.validate();
```
Commented-out code. Just delete.

cheqianh replied:
Deleted.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -265,115 +266,111 @@ private static void processToEventFromIonStream(IonWriter ionWriterForOutput,
+                            imports[i].getImportName(), imports[i].getVersion(), null);
+                    symbolTables[i] = symbolTable;
+                }
+                ionWriterForOutput = args.getOutputFormat().createIonWriterWithImports(outputStream, symbolTables);
```
This reassignment does not last beyond the return of this method. You probably need to add the IonWriter to the ProcessContext so you can reassign it there and keep the new writer in scope for data written after the return of this method.

cheqianh replied:
Changed.

I Create a ionWriter at IonProcess.java 76 and close it at IonProcess.java 81-90

I did same thing for embedded Stream since we may face symbol table within an embedded stream.
(I create the processContext for embeddedStream at Ionprocess.java 340 and close it at line 375.)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -61,66 +62,68 @@ public static void main(final String[] args) {
         }
 
+        //Use this variable for symbol_table event which changes ionWriter
+        IonWriter ionWriterForOutput = null;
```
See my commend below about adding the IonWriter to the ProcessContext. Currently this doesn't have any effect. You'll probably also have to remove the IonWriter instantiation from the try-with-resources and add a `finally` block in which you manually close it, since the reference may change during execution.

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -392,15 +389,22 @@ private static void embeddedEventToIon(IonReader ionReader,
+                                    imports[i].getImportName(), imports[i].getVersion(), null);
+                            symbolTables[i] = symbolTable;
+                        }
+                        ionWriterForOutput = args.getOutputFormat()
```
Same comment about reassigning an IonWriter in the ProcessContext.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -265,115 +266,111 @@ private static void processToEventFromIonStream(IonWriter ionWriterForOutput,
+            } else if (event.getEventType().equals(EventType.SCALAR)) {
+                writeIonByType(event, ionWriterForOutput);
+            } else if (event.getEventType().equals(EventType.SYMBOL_TABLE)) {
+                ionWriterForOutput.finish();
```
This should be `close()`, which finishes the writer and frees any resources it consumes.

cheqianh replied:
Changed

question: 

If I finish the writer and free outputStream of this writer, the outputStream will be closed so my new ionWriter can't write to the same outputStream anymore. Did I miss anything?

tgregg replied:
You can create a new OutputStream that appends to the existing file. See https://docs.oracle.com/javase/7/docs/api/java/io/FileOutputStream.html#FileOutputStream(java.io.File,%20boolean)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -392,15 +389,22 @@ private static void embeddedEventToIon(IonReader ionReader,
                         tempWriter.stepOut();
                     } else if (event.getEventType() == EventType.SYMBOL_TABLE) {
-                        //TODO symbol table
+                        ionWriterForOutput.finish();
```
This should be `close()`, which finishes the writer and frees any resources it consumes.

cheqianh replied:
Changed

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -392,15 +389,22 @@ private static void embeddedEventToIon(IonReader ionReader,
                     } else if (event.getEventType() == EventType.SYMBOL_TABLE) {
-                        //TODO symbol table
+                        ionWriterForOutput.finish();
+                        ImportDescriptor[] imports = event.getImports();
+                        SymbolTable[] symbolTables = new SymbolTable[imports.length];
+                        for (int i = 0; i < imports.length; i++) {
+                            SymbolTable symbolTable = IonSystemBuilder.standard().build().newSharedSymbolTable(
+                                    imports[i].getImportName(), imports[i].getVersion(), null);
+                            symbolTables[i] = symbolTable;
+                        }
+                        ionWriterForOutput = args.getOutputFormat()
```
Looks like this code block is duplicated in processToNormalFromEventStream. Looks like a good candidate for its own method.

cheqianh replied:
Combined (ionProcess.java 390)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -474,7 +436,7 @@ private static Event ionStreamToEvent(IonReader ionReader) {
                 case "field_name":
                     ionReader.stepIn();
-                    String fieldText = null;
+                    String fieldText = "";
```
As discussed in a different comment, you should be able to create a SymbolToken with `null` text. This is actually the correct thing to do, since an empty string is a valid field name.

cheqianh replied:
Changed, it performances as ion-c now.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -544,14 +502,28 @@ private static Event ionStreamToEvent(IonReader ionReader) {
 
+        }
+        //compare text value and binary value
+        if (textValue != null && binaryValue != null) {
```
If this branch isn't taken, throw an exception, as the event stream is illegal.

cheqianh replied:
Wrote a valid function to check each cases.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -544,14 +502,28 @@ private static Event ionStreamToEvent(IonReader ionReader) {
+        }
+        //compare text value and binary value
+        if (textValue != null && binaryValue != null) {
+            IonSystem ionSystem = IonSystemBuilder.standard().build();
```
This can be a constant. No need to reinstantiate it every time.

cheqianh replied:
Changed (IonProcess.java 47)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -544,14 +502,28 @@ private static Event ionStreamToEvent(IonReader ionReader) {
+            if (!Equivalence.ionEquals(text, binary)) {
+                throw new IonException("invalid Event: Text value and Binary value are different");
+            } else {
+                IonValue v = IonSystemBuilder.standard().build().singleValue(textValue);
```
No need to re-read it here; you've already got this value stored in `text` above.

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -544,14 +502,28 @@ private static Event ionStreamToEvent(IonReader ionReader) {
+
     private static ImportDescriptor[] ionStreamToImportDescriptors(IonReader ionReader) {
-        LinkedList<ImportDescriptor> imports = new LinkedList<>();
+        ArrayList<ImportDescriptor> imports = new ArrayList();
```
When you're assigning to a list, just use `List` as the declared type unless there is something you absolutely need from the particular `List` implementation. That way if you change implementations you only have to make a change in one place. Also, always use `<>` to indicate that the type you're instantiating matches what was declared.

```java
List<ImportDescriptor> imports = new ArrayList<>();
```

cheqianh replied:
Changed.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -630,37 +593,33 @@ private static Event ionStreamToEvent(EventType eventType,
-                    ByteArrayOutputStream binaryOut = new ByteArrayOutputStream();
-                    IonWriter binaryWriter = IonBinaryWriterBuilder.standard().build(binaryOut);
-                    ) {
+                    IonWriter tempWriter = IonTextWriterBuilder.standard().build(textOut);
```
One thing we can do is create a constant `IonTextWriterBuilder` instance and reuse it whenever we need a temporary text writer. The builder just holds configuration, so it's reusable. This will improve performance.

cheqianh replied:
Added.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -630,37 +593,33 @@ private static Event ionStreamToEvent(EventType eventType,
+        IonValue value = null;
         if (eventType == EventType.SCALAR) {
             try (
                     ByteArrayOutputStream textOut = new ByteArrayOutputStream();
```
Might as well use a StringBuilder here too.

cheqianh replied:
Changed. (682)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -630,37 +593,33 @@ private static Event ionStreamToEvent(EventType eventType,
         else if (newTable == null) return false;
-
-        if (newTable.isLocalTable() && newTable.getImportedTables().length == 0) {
+        if (newTable.isSystemTable() && newTable.getName() == "$ion") {
```
Why isn't this `if (newTable.isSystemTable() && newTable.isSystemTable())`?

cheqianh replied:
Changed.

question:
what version of system table are we using? it's 0 right?

cheqianh committed:
validation and final variables

cheqianh committed:
only have reassign ionWriter part

cheqianh committed:
added new ionWriter to processContext

cheqianh committed:
finish the processcontext ionwriter part

cheqianh committed:
finish validate

cheqianh committed:
checked logic

cheqianh committed:
finish style check

cheqianh committed:
ready for cr

cheqianh committed:
little change

cheqianh commented:
Sorry, forgot to squash the commits. 

The new commits starts from **validation and final variables** (912bfc60)

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -62,67 +63,64 @@ public static void main(final String[] args) {
-                            .writeOutput(ionWriterForErrorReport);
-                }
+                new ErrorDescription(ErrorType.STATE, e.getMessage(), processContext.getFileName(),
+                        processContext.getEventIndex()).writeOutput(ionWriterForErrorReport);
```
Looks like you can probably combine these two `catch` branches now.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -305,78 +300,57 @@ private static void processToNormalFromEventStream(IonWriter ionWriterForOutput,
+            if (event.getEventType() == EventType.CONTAINER_START) {
                 if (isEmbeddedEvent(event)) {
-                    embeddedEventToIon(ionReader, ionWriterForOutput, processContext, args, outputStream);
+                    embeddedEventToIon(ionReader,processContext, args, outputStream);
```
Missing a space after comma.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -387,57 +361,101 @@ private static void embeddedEventToIon(IonReader ionReader,
+                    try {
+                        ionWriter.close();
+                    } catch (IOException e) {
+                        System.err.println(e.getMessage());
```
You can't keep going after this, right? I think this exception should either be allowed to bubble up, or you should exit from here.

tgregg commented on test/tools/cli/IonProcess.java:
```diff
@@ -387,57 +361,101 @@ private static void embeddedEventToIon(IonReader ionReader,
+        if (field != null) {
+            if (!ionWriter.isInStruct()) throw new IonException("invalid field_name inside STRUCT");
+            ionWriter.setFieldNameSymbol(field);
+        } else if (ionWriter.isInStruct()) {
```
Should this be an error? Inside a struct, every value *must* have a field name symbol token.

tgregg reviewed (approved):
Great! These comments are minor. Feel free to fix them and merge without creating another revision of this PR.

cheqianh commented:
Thanks for the code review, I'll fix them in the next commit!
